velox/expression/tests/SignatureTemplateBinder.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/tests/SignatureTemplateBinder.h"
+
+namespace facebook::velox::test {
+
+namespace {
+
+// Return a random type among those in typeList determined by seed.
+TypePtr randomType(std::mt19937& seed) {
+  static std::vector<TypePtr> typeList{
+      BOOLEAN(), TINYINT(), SMALLINT(), INTEGER(), BIGINT(), REAL(), DOUBLE()};
+  auto index = folly::Random::rand32(typeList.size(), seed);
+  return typeList[index];
+}
+
+std::string fromTypeToBaseName(const TypePtr& type) {
+  std::string typeName = type->kindName();
+  std::transform(
+      typeName.begin(), typeName.end(), typeName.begin(), [](unsigned char c) {
+        return std::tolower(c);
+      });
+  return typeName;
+}
+
+std::optional<TypeKind> fromBaseNameToTypeKind(std::string& typeName) {
+  std::transform(
+      typeName.begin(), typeName.end(), typeName.begin(), [](unsigned char c) {
+        return std::toupper(c);
+      });
+  return tryMapNameToTypeKind(typeName);
+}
+
+} // namespace
+
+// Return a concrete type for formalType under the constraints in bindings. If
+// formalType contains type variables not bound by bindings, bing this type
+// variable to a random type.
+TypePtr SignatureTemplateBinder::bindPossibleType(
+    const exec::TypeSignature& formalType) {
+  // Not supporting argument types that contain "any".
+  std::string type = formalType.baseType();
+  if (type == "any") {
+    VELOX_NYI();
+  }
+
+  auto it = bindings_.find(type);
+  if (it != bindings_.end()) {
+    if (!it->second) {
+      it->second = randomType(seed_);
+    }
+    VELOX_DCHECK(it->second);
+    return it->second;
+  } else {
+    auto typeKind = fromBaseNameToTypeKind(type);
+    if (!typeKind.has_value()) {
+      return nullptr;
+    }
+
+    if (typeKind == TypeKind::ARRAY) {
+      auto child = bindPossibleType(formalType.parameters()[0]);
+      if (!child) {
+        return nullptr;
+      }
+      return ARRAY(std::move(child));
+
+    } else if (typeKind == TypeKind::MAP) {
+      auto key = bindPossibleType(formalType.parameters()[0]);
```

Comment:
The logic for resolving child types can be unified across ARRAY, MAP, ROW and custom complex types. We can loop over formalType.parameters() and build `std::vector<TypePtr> children;`, then create type using type name and child types. See SignatureBinder::tryResolveType logic for an example.